src/library_fs.js

```diff
@@ -985,20 +1005,37 @@ FS.staticInit();
     },
     fchown(fd, uid, gid) {
       var stream = FS.getStreamChecked(fd);
-      FS.chown(stream.node, uid, gid);
+      var node = stream.node;
+      var attrs = {
+        timestamp: Date.now()
+        // we ignore the uid / gid for now
+      };
+      if (stream.stream_ops.getattr) {
+        return stream.stream_ops.setattr(stream, attrs);
+      } else if (node.node_ops.getattr) {
+        return node.node_ops.setattr(node, attrs);
+      }
+      throw new FS.ErrnoError({{{ cDefs.EPERM }}});
     },
-    truncate(path, len) {
+    truncateCommon(arg, len, ftruncate) {
       if (len < 0) {
         throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
       }
       var node;
-      if (typeof path == 'string') {
-        var lookup = FS.lookupPath(path, { follow: true });
+      var stream;
+      if (ftruncate) {
+        stream = FS.getStreamChecked(arg);
+        if ((stream.flags & {{{ cDefs.O_ACCMODE }}}) === {{{ cDefs.O_RDONLY}}}) {
+          throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
+        }  
+        node = stream.node;
+      } else if (typeof arg == 'string') {
+        var lookup = FS.lookupPath(arg, { follow: true });
         node = lookup.node;
       } else {
-        node = path;
+        node = arg;
       }
```

Comment:
Can we leave lines 1025 through 1037 in the caller and also avoid having a polymorphic `arg` here?

How about `truncateCommon(node, stream, len)` here node would be required and stream would be optional.

Reply:
I'm not entirely clear what you mean by this but I did something in [989cf22](https://github.com/emscripten-core/emscripten/pull/23058/commits/989cf22b8aa5080ad0ec2f2dfa76b2eeac1cbbb9), let me know if that's not what you were suggesting.